src/Npgsql/pg_catalog/dat2json.pl

```diff
@@ -0,0 +1,276 @@
+use constant PG_RANGE_KEYS => qw(rngtypid rngsubtype rngmultitypid);
+use constant PG_TYPE_KEYS => qw(oid typname typtype array_type_oid);
+
+die "Please pass the access token" unless defined $ARGV[0];
+run();
+
+sub run {
+    my $access_token = shift(@ARGV);
+    my ($pg_range, $pg_type, $latestMajor, $latestBranch, $pg_range_url, $pg_type_url, $httpResponse);
+    my $ua = LWP::UserAgent->new;
+    $ua->agent("Npgsql catalog updater");
+
+    # Try to get the .dat files latest stable PostgreSQL release by trying urls while incrementing the major version
+    # starting at LATEST_KNOWN_POSTGRES_MAJOR
+    for (my $i = LATEST_KNOWN_POSTGRES_MAJOR;$i < 100;$i++) {
+        my $currentBranch = sprintf(BRANCH_TEMPLATE, $i);
+        $pg_range_url = sprintf(URL_TEMPLATE, PG_RANGE_CATALOG_FILE_NAME, $currentBranch);
+        my $req = HTTP::Request->new(GET => $pg_range_url);
+        $httpResponse = $ua->request($req);
+        last unless $httpResponse->is_success;
+        $pg_range = $httpResponse->decoded_content;
+
+        $pg_type_url = sprintf(URL_TEMPLATE, PG_TYPE_CATALOG_FILE_NAME, $currentBranch);
+        $req = HTTP::Request->new(GET => $pg_type_url);
+        $httpResponse = $ua->request($req);
+        last unless $httpResponse->is_success;
+        $pg_type = $httpResponse->decoded_content;
+        $latestMajor = $i;
+        $latestBranch = $currentBranch;
+    }
+
+    die "Failed to get ${\PG_RANGE_CATALOG_FILE_NAME} from $pg_range_url: ${\$httpResponse->status_line}" unless defined $pg_range;
+    die "Failed to get ${\PG_TYPE_CATALOG_FILE_NAME} from $pg_type_url: ${\$httpResponse->status_line}" unless defined $pg_type;
+
+    $pg_type = filter_dat_entries($pg_type, PG_TYPE_KEYS);
+    $pg_range = filter_dat_entries($pg_range, PG_RANGE_KEYS);
+
+    my $json = JSON->new->allow_nonref;
+    $json->canonical();
+
+    my $newJson = $json->pretty->encode( { pg_range => $pg_range, pg_type => $pg_type, version => $latestMajor } );
+    my $catalogs_json_path = sprintf("%s/catalogs.json", dirname(__FILE__));
+    open(my $fh, '+<:raw', $catalogs_json_path) or die "Failed to open '$catalogs_json_path': $!";
+    read $fh, my $oldJson, -s $fh;
+    if ($newJson ne $oldJson) {
+        print "Detected changes in $latestBranch.\n";
+        truncate $fh, 0;
+        seek $fh, 0, 0;
+        print $fh $newJson;
+        commit_and_pr($access_token, $latestBranch, encode_base64($newJson, ''));
+    }
+    else {
+        print "The file '$catalogs_json_path' is up to date.\n";
+    }
+    close($fh);
+}
+
+sub filter_dat_entries {
+    my $datArray = eval(shift);
+    my @wantedKeys = @_;
+
+    for my $i (0 .. $#$datArray) {
+        # We don't use hash slicing here since this adds missing keys with null values
+        foreach my $key (keys %{${$datArray}[$i]}) {
+            delete ${${$datArray}[$i]}{$key} unless (grep $_ eq $key, @wantedKeys);
+        }
+    }
+    return $datArray;
+}
+
+sub commit_and_pr {
```

Comment:
One additional thought I had... Assuming you want to keep doing the PR automatically, it's probably easier to do that in the Github Actions workflow rather than in Perl, with explicit GraphQL calls and all. There's the [Github CLI](https://cli.github.com), which is supposed to make this kind of stuff very easy (and I assume is available on the Github Actions images, and hopefully integrates well with the env vars and everything). This would also allow us to easily run the script locally/for testing without generating a PR etc.

Reply:
I'll have a look at this.

Reply:
It turns out that this is still a multi-step procedure which would only involve the GitHub CLI to create the PR and multiple plain Git commands (branch, add, commit) before where I'd have to convey information via environment variables and use process exit codes.
I'd argue that doing this in BASH is not really better than doing it in Perl - at least to me it feels a lot worse.

> This would also allow us to easily run the script locally/for testing without generating a PR etc.

I've found a very easy solution for this: Just don't pass the access token and the script will skip the PR part.